test/google_search_data_viewer_web/controllers/keyword_controller_test.exs

```diff
@@ -0,0 +1,35 @@
+defmodule GoogleSearchDataViewerWeb.KeywordControllerTest do
+  use GoogleSearchDataViewerWeb.ConnCase, async: true
+
+  describe "GET /keywords" do
+    test "renders keywords page", %{conn: conn} do
+      user = insert(:user)
+
+      conn =
+        conn
+        |> init_test_session(user_id: user.id)
+        |> get(Routes.keyword_path(conn, :index))
+
+      assert html_response(conn, 200) =~ "Select a csv file to upload keywords"
+    end
+  end
+
+  describe "POST /keywords/upload" do
+    test "given a valid csv file extention, uploads the file", %{conn: conn} do
+      file = %Plug.Upload{
+        path: "test/support/fixtures/keywords/valid_keywords.csv",
+        filename: "valid_keywords.csv"
+      }
+
+      user = insert(:user)
+
+      conn =
+        conn
+        |> init_test_session(user_id: user.id)
+        |> post("/keywords/upload", %{:file => file})
+
+      assert redirected_to(conn, 302) =~ "/keywords"
+      assert get_flash(conn, :info) =~ "File successfully uploaded"
+    end
```

Comment:
Here you can add one more test for uploading by not logged in user.

Reply:
> 

Ok, will do. I initially thought I would carry out those tests in the `ensure_authenticated` plug test but I will add that here too.